Objects/typeobject.c

```diff
@@ -7000,6 +7005,34 @@ static slotdef slotdefs[] = {
     {NULL}
 };
 
+
+/* Since searching slotdefs[] is expensive and assigning to slots is rare,
+   do a quick probability check if the PyUnicode name could be a slot name
+   at all. */
+static int name_looks_like_slot(PyObject *name) {
+    Py_ssize_t length = PyUnicode_GET_LENGTH(name);
+    /* The longest slot name is currently "__getattribute__": 16 characters */
+    if (length >= 6 && length <= 16) {
+        /* All slot names are pure ASCII. */
+        int kind = PyUnicode_KIND(name);
+        if (kind == PyUnicode_1BYTE_KIND) {
+            void *data = PyUnicode_DATA(name);
+            /* Allow the C compiler to read 2 bytes instead of two times 1 byte. */
+            if ((PyUnicode_READ(kind, data, 0) == '_') &
```

Comment:
Also, to make this a bit more clear, could we invert every check and do early returns?

I think that makes the function more clear and reduces a lot the indentation levels.

Reply:
I would just merge two ifs.

Parenthesis are not needed around the `==` operator. IMHO they reduce readability.

Reply:
For me, they increase readability, because they make the separate conditions explicit. Also, C compilers often warn about "&" in plain boolean expressions, because its usage can hint at a typo.
I'll reduce the nesting, though.

Reply:
Is not there a typo? Perhaps you meant "&&"?

Reply:
Read the comment above the line. :)
When I write `a[0] == 1 && a[1] == 2 && a[2] == 3 && ...`, then the C compiler has to assume that there is a data dependency between the three items (`a[1]` might only be readable if `a[0] == 1`, and might otherwise segfault – think of 0-terminated C strings), so it must generate slow and branchy code that reads one byte at a time.
When, instead, I write `(a[0] == 1) & (a[1] == 2) & (a[2] == 3) & (a[3] == 4) & ...`, then the C compiler can safely assume that all fields are accessible and that there will be no segfault, so that it can generate a single memory access and comparison for it, e.g. reading ints instead of bytes.
Basically, I'm leaving it to the C compiler to do the right thing, instead of preventing it from taking the best decision.

Reply:
>Read the comment above the line. :)
When I write `a[0] == 1 && a[1] == 2 && a[2] == 3 && ...`, then the C compiler has to assume that there is a data dependency between the three items (`a[1]` might only be readable if `a[0] == 1`, and might otherwise segfault – think of 0-terminated C strings), so it must generate slow and branchy code that reads one byte at a time.
When, instead, I write `(a[0] == 1) & (a[1] == 2) & (a[2] == 3) & (a[3] == 4) & ...`, then the C compiler can safely assume that all fields are accessible and that there will be no segfault, so that it can generate a single memory access and comparison for it, e.g. reading ints instead of bytes.
Basically, I'm leaving it to the C compiler to do the right thing, instead of preventing it from taking the best decision.

I am not very sure about this assertion. This is the x86-64 generated for the two options:

https://godbolt.org/z/YfgzAi

Unless I am missing something, both generate the same branches and I would even say that the 2nd version generates more efficient asm (this probably will be very susceptible to other aspects of the surrounding code so I don't take this very seriously :) ). I would also want to run some benchmarks to check if there is any difference worth the surprise when reading & instead of &&.

I know it sounds cliché but in general, is best not the help the compiler. This is specially true when running with very different set of compilers/platforms.

Reply:
I played around with the code in godbolt and I think I found a reason for this. gcc seems to consider a word access more costly than a byte access. You can see that when you cast the byte pointer to a short pointer to compare the first 16 bits in one go. It then moves the start word access after the end byte access when allowed by the "&".

Anyway, I see the argument of least surprise. I'll use "&&". Thanks for insisting.